src/main/java/com/launchdarkly/client/RedisFeatureStore.java

```diff
@@ -245,26 +274,64 @@ public boolean initialized() {
     return getInit();
   }
 
+  /**
+   * Releases all resources associated with the store. The store must no longer be used once closed.
+   * @throws IOException
+   */
+  public void close() throws IOException
+  {
+    pool.destroy();
+  }
+
+
 
   private String featuresKey() {
     return prefix + ":features";
   }
 
   private Boolean getInit() {
-    return jedis.exists(featuresKey());
+    Jedis jedis = null;
+
+    try {
+      jedis = getJedis();
+      return jedis.exists(featuresKey());
+    } finally {
+      if (jedis != null) {
+        jedis.close();
+      }
+    }
   }
 
   private FeatureRep<?> getRedis(String key) {
-    Gson gson = new Gson();
-    String featureJson = jedis.hget(featuresKey(), key);
+    Jedis jedis = null;
+    try {
+      jedis = getJedis();
+      Gson gson = new Gson();
+      String featureJson = getJedis().hget(featuresKey(), key);
+
+      if (featureJson == null) {
+        return null;
+      }
+
+      Type type = new TypeToken<FeatureRep<?>>() {}.getType();
+      FeatureRep<?> f = gson.fromJson(featureJson, type);
 
-    if (featureJson == null) {
-      return null;
+      return f.deleted ? null : f;
+    } finally {
+      if (jedis != null) {
+        jedis.close();
+      }
     }
+  }
 
-    Type type = new TypeToken<FeatureRep<?>>() {}.getType();
-    FeatureRep<?> f = gson.fromJson(featureJson, type);
+  private final Jedis getJedis() {
+    return pool.getResource();
+  }
 
-    return f.deleted ? null : f;
+  private final JedisPoolConfig getPoolConfig() {
+    JedisPoolConfig config = new JedisPoolConfig();
+    config.setMaxTotal(256);
+    config.setBlockWhenExhausted(false);
```

Comment:
Due to the nature of the store abstraction, blocking when exhausted can cause a deadlock. This can happen because a single toggle call can require multiple jedis pool connections to complete-- one to check whether the store is initialized, and another to fetch the value from the store. If multiple threads attempt to do this concurrently, and the pool is near exhaustion, the could both acquire connections to check initialization, and then block waiting for another connection to fetch the store value.